File: examples/config-transfer/go/config_transfer.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"

	"golang.org/x/crypto/argon2"
)

const (
	encryptedFormat = "yinshu-config-encrypted"
	payloadFormat   = "yinshu-config"
	version         = 1
	saltBytes       = 16
	nonceBytes      = 12
	keyBytes        = 32
	memoryKiB       = 19456
	iterations      = 2
	parallelism     = 1
	tagBytes        = 16
	expectedPayload = "OTX3vkYug76bv335qmWdp85pbgu85QfwarlnqhxGoV0U+4sRez0dlwWy+5eIe597KLRqdHg7XJVbjLds/mXROcLHLhTJrJJ+DWpB2Xc6BX2sKii+bziOsb8akhUwxqo="
)

type EncryptOptions struct {
	Salt  []byte
	Nonce []byte
}

type EncryptedConfigFile struct {
	Format  string         `json:"format"`
	Version int            `json:"version"`
	Crypto  CryptoMetadata `json:"crypto"`
	Payload string         `json:"payload"`
}

type CryptoMetadata struct {
	KDF         string `json:"kdf"`
	MemoryKiB   uint32 `json:"memory_kib"`
	Iterations  uint32 `json:"iterations"`
	Parallelism uint8  `json:"parallelism"`
	Cipher      string `json:"cipher"`
	TagBytes    int    `json:"tag_bytes"`
	Salt        string `json:"salt"`
	Nonce       string `json:"nonce"`
}

type transferPayload struct {
	Format  string `json:"format"`
	Version int    `json:"version"`
	Config  struct {
		Service struct {
			Port int `json:"port"`
		} `json:"service"`
	} `json:"config"`
}

func EncryptConfigFile(payload any, password string, options EncryptOptions) (*EncryptedConfigFile, error) {
	salt, err := fixedOrRandom(options.Salt, saltBytes)
	if err != nil {
		return nil, err
	}
	nonce, err := fixedOrRandom(options.Nonce, nonceBytes)
	if err != nil {
		return nil, err
	}

	key := argon2.IDKey([]byte(password), salt, iterations, memoryKiB, parallelism, keyBytes)
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCMWithTagSize(block, tagBytes)
	if err != nil {
		return nil, err
	}

	plaintext, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}
	encryptedPayload := gcm.Seal(nil, nonce, plaintext, nil)

	return &EncryptedConfigFile{
		Format:  encryptedFormat,
		Version: version,
		Crypto: CryptoMetadata{
			KDF:         "argon2id13",
			MemoryKiB:   memoryKiB,
			Iterations:  iterations,
			Parallelism: parallelism,
			Cipher:      "aes-256-gcm",
			TagBytes:    tagBytes,
			Salt:        base64.StdEncoding.EncodeToString(salt),
			Nonce:       base64.StdEncoding.EncodeToString(nonce),
		},
		Payload: base64.StdEncoding.EncodeToString(encryptedPayload),
	}, nil
}

func DecryptConfigFile(file EncryptedConfigFile, password string) (map[string]any, error) {
	if err := validateEnvelope(file); err != nil {
		return nil, err
	}

	salt, err := decodeFixed(file.Crypto.Salt, saltBytes)
	if err != nil {
		return nil, err
	}
	nonce, err := decodeFixed(file.Crypto.Nonce, nonceBytes)
	if err != nil {
		return nil, err
	}
	encryptedPayload, err := base64.StdEncoding.DecodeString(file.Payload)
	if err != nil {
		return nil, err
	}

	key := argon2.IDKey([]byte(password), salt, iterations, memoryKiB, parallelism, keyBytes)
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCMWithTagSize(block, tagBytes)
	if err != nil {
		return nil, err
	}
	plaintext, err := gcm.Open(nil, nonce, encryptedPayload, nil)
	if err != nil {
		return nil, errors.New("invalid password or encrypted config payload")
	}

	var payload map[string]any
	if err := json.Unmarshal(plaintext, &payload); err != nil {
		return nil, err
	}
	if payload["format"] != payloadFormat || payload["version"] != float64(version) {
		return nil, errors.New("invalid decrypted config payload")
	}
	return payload, nil
}

func fixedOrRandom(value []byte, length int) ([]byte, error) {
	if value != nil {
		if len(value) != length {
			return nil, errors.New("invalid salt or nonce length")
		}
		return value, nil
	}

	out := make([]byte, length)
	if _, err := io.ReadFull(rand.Reader, out); err != nil {
		return nil, err
	}
	return out, nil
}

func validateEnvelope(file EncryptedConfigFile) error {
	if file.Format != encryptedFormat ||
		file.Version != version ||
		file.Crypto.KDF != "argon2id13" ||
		file.Crypto.MemoryKiB != memoryKiB ||
		file.Crypto.Iterations != iterations ||
		file.Crypto.Parallelism != parallelism ||
		file.Crypto.Cipher != "aes-256-gcm" ||
		file.Crypto.TagBytes != tagBytes {
		return errors.New("invalid encrypted config file")
	}
	return nil
}

func decodeFixed(value string, length int) ([]byte, error) {
	bytes, err := base64.StdEncoding.DecodeString(value)
	if err != nil {
		return nil, err
	}
	if len(bytes) != length {
		return nil, errors.New("invalid fixed-length field")
	}
	return bytes, nil
}

func verify() error {
	payload := transferPayload{
		Format:  payloadFormat,
		Version: version,
	}
	payload.Config.Service.Port = 17890

	salt, _ := hex.DecodeString("000102030405060708090a0b0c0d0e0f")
	nonce, _ := hex.DecodeString("101112131415161718191a1b")
	file, err := EncryptConfigFile(payload, "test-password", EncryptOptions{Salt: salt, Nonce: nonce})
	if err != nil {
		return err
	}
	if file.Payload != expectedPayload {
		return errors.New("encrypted payload does not match expected vector")
	}

	decrypted, err := DecryptConfigFile(*file, "test-password")
	if err != nil {
		return err
	}
	config := decrypted["config"].(map[string]any)
	service := config["service"].(map[string]any)
	if service["port"] != float64(17890) {
		return errors.New("decrypted payload did not round-trip")
	}

	encoded, err := json.MarshalIndent(file, "", "  ")
	if err != nil {
		return err
	}
	fmt.Println(string(encoded))
	return nil
}

func main() {
	if len(os.Args) > 1 && os.Args[1] == "verify" {
		if err := verify(); err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
	}
}

File: examples/config-transfer/node/config-transfer.mjs
import crypto from 'node:crypto';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
const sodium = require('libsodium-wrappers-sumo');

const ENCRYPTED_FORMAT = 'yinshu-config-encrypted';
const PAYLOAD_FORMAT = 'yinshu-config';
const VERSION = 1;
const SALT_BYTES = 16;
const NONCE_BYTES = 12;
const KEY_BYTES = 32;
const MEMORY_KIB = 19456;
const ITERATIONS = 2;
const PARALLELISM = 1;
const TAG_BYTES = 16;
const EXPECTED_PAYLOAD =
  'OTX3vkYug76bv335qmWdp85pbgu85QfwarlnqhxGoV0U+4sRez0dlwWy+5eIe597KLRqdHg7XJVbjLds/mXROcLHLhTJrJJ+DWpB2Xc6BX2sKii+bziOsb8akhUwxqo=';

await sodium.ready;

export function encryptConfigFile(payload, password, options = {}) {
  const salt = fixedOrRandom(options.salt, SALT_BYTES);
  const nonce = fixedOrRandom(options.nonce, NONCE_BYTES);
  const key = deriveKey(password, salt);
  const cipher = crypto.createCipheriv('aes-256-gcm', key, nonce, {
    authTagLength: TAG_BYTES,
  });
  const ciphertext = Buffer.concat([
    cipher.update(JSON.stringify(payload), 'utf8'),
    cipher.final(),
  ]);
  const tag = cipher.getAuthTag();

  return {
    format: ENCRYPTED_FORMAT,
    version: VERSION,
    crypto: {
      kdf: 'argon2id13',
      memory_kib: MEMORY_KIB,
      iterations: ITERATIONS,
      parallelism: PARALLELISM,
      cipher: 'aes-256-gcm',
      tag_bytes: TAG_BYTES,
      salt: Buffer.from(salt).toString('base64'),
      nonce: Buffer.from(nonce).toString('base64'),
    },
    payload: Buffer.concat([ciphertext, tag]).toString('base64'),
  };
}

export function decryptConfigFile(file, password) {
  validateEnvelope(file);

  const salt = decodeFixed(file.crypto.salt, SALT_BYTES);
  const nonce = decodeFixed(file.crypto.nonce, NONCE_BYTES);
  const encryptedPayload = Buffer.from(file.payload, 'base64');
  if (encryptedPayload.length < TAG_BYTES) {
    throw new Error('Invalid encrypted config payload.');
  }

  const ciphertext = encryptedPayload.subarray(0, -TAG_BYTES);
  const tag = encryptedPayload.subarray(-TAG_BYTES);
  const decipher = crypto.createDecipheriv('aes-256-gcm', deriveKey(password, salt), nonce, {
    authTagLength: TAG_BYTES,
  });
  decipher.setAuthTag(tag);

  let plaintext;
  try {
    plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString('utf8');
  } catch {
    throw new Error('Invalid password or encrypted config payload.');
  }

  const payload = JSON.parse(plaintext);
  if (payload.format !== PAYLOAD_FORMAT || payload.version !== VERSION) {
    throw new Error('Invalid decrypted config payload.');
  }
  return payload;
}

function deriveKey(password, salt) {
  return Buffer.from(
    sodium.crypto_pwhash(
      KEY_BYTES,
      password,
      salt,
      ITERATIONS,
      MEMORY_KIB * 1024,
      sodium.crypto_pwhash_ALG_ARGON2ID13,
    ),
  );
}

function fixedOrRandom(value, length) {
  if (value) {
    const bytes = Buffer.from(value);
    if (bytes.length !== length) {
      throw new Error('Invalid salt or nonce length.');
    }
    return bytes;
  }
  return crypto.randomBytes(length);
}

function decodeFixed(value, length) {
  const bytes = Buffer.from(value, 'base64');
  if (bytes.length !== length) {
    throw new Error('Invalid fixed-length field.');
  }
  return bytes;
}

function validateEnvelope(file) {
  if (
    file?.format !== ENCRYPTED_FORMAT ||
    file.version !== VERSION ||
    file.crypto?.kdf !== 'argon2id13' ||
    file.crypto.memory_kib !== MEMORY_KIB ||
    file.crypto.iterations !== ITERATIONS ||
    file.crypto.parallelism !== PARALLELISM ||
    file.crypto.cipher !== 'aes-256-gcm' ||
    file.crypto.tag_bytes !== TAG_BYTES
  ) {
    throw new Error('Invalid encrypted config file.');
  }
}

function verify() {
  const payload = {
    format: PAYLOAD_FORMAT,
    version: VERSION,
    config: {
      service: {
        port: 17890,
      },
    },
  };
  const file = encryptConfigFile(payload, 'test-password', {
    salt: Buffer.from('000102030405060708090a0b0c0d0e0f', 'hex'),
    nonce: Buffer.from('101112131415161718191a1b', 'hex'),
  });

  if (file.payload !== EXPECTED_PAYLOAD) {
    throw new Error('Encrypted payload does not match expected vector.');
  }

  const decrypted = decryptConfigFile(file, 'test-password');
  if (decrypted.config.service.port !== 17890) {
    throw new Error('Decrypted payload did not round-trip.');
  }

  console.log(JSON.stringify(file, null, 2));
}

if (process.argv[2] === 'verify') {
  verify();
}
